src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py

```diff
@@ -921,14 +921,34 @@ def _resolve_policy_id(policy, policy_set_definition, client):
     policy_id = policy or policy_set_definition
     if not is_valid_resource_id(policy_id):
         if policy:
-            policy_def = client.policy_definitions.get(policy)
+            policy_def = _get_custom_or_builtin_policy(client, policy)
             policy_id = policy_def.id
         else:
-            policy_set_def = client.policy_set_definitions.get(policy_set_definition)
+            policy_set_def = _get_custom_or_builtin_policy(client, policy_set_definition, 'setdefinition')
             policy_id = policy_set_def.id
     return policy_id
 
 
+def _get_custom_or_builtin_policy(client, name, resourcetype='definition'):
+    if resourcetype != 'definition':
+        if supported_api_version(ResourceType.MGMT_RESOURCE_POLICY, min_api='2017-06-01-preview'):
+            errorObject = get_sdk(ResourceType.MGMT_RESOURCE_POLICY, 'ErrorResponseException', mod='models')
+            try:
+                return client.policy_set_definitions.get(name)
+            except errorObject as ex:
+                if ex.response.status_code == 404:
```

Comment:
If the status code here is not 404, you end up with an exception thrown on line 941. The exception message doesn't fit the description of what actually happens.

Reply:
The exception message on 941 is thrown if the supported api version in L934 fails, not when the status code is anything other than 404.

Reply:
Then your code flow doesn't reflect your intention.
1. Line 938, exception captured.
2. Line 939, test fails when status_code is 500
3. Line 940 is skipped and line 941 is run, therefore exception indicating profile mismatch is thrown.

Reply:
You need to test all the branches of the condition flow here since it is clear to me that the error handling is not tested.

Additionally, if you raise an exception here, the method's error handling mechanism is inconsistent because, on line 949, the non-404 `CloudError` causes a `None` to be returned.

Reply:
I don't think we need to test all branches of the condition flow. I'm actually only moving [a part of the code](https://github.com/Azure/azure-cli/blob/master/src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py#L979) we already had to a new function.

Reply:
The fact that in the code review I'm managed to point out multiple issues in this condition structure and exception re-raise is added in following commit, I think you need to carefully test it. Code is changed during movement.

Reply:
What do you mean when you say "carefully test it"? As in, add scenario tests? How are we supposed to mimic a 500 response, if that is the case?

From the code stand point, compared to what we already have in place, there hasn't been a major change. The flow is:
```
1. Get policy
2. Catch exception, if it is 404, get built in policy (test already included for this)
3. If other exception, raise
```
This is the exact same workflow that is in place already, like I pointed out, the only difference being, instead of hand-constructing a URL for built in policy GET, we invoke an SDK call.

Reply:
@vivsriaus, it is great to see the new API to get the `built-in-policies`! Here are my 2 cents to consolidate the code:
1. Can we remove the `supported_api_version` which fuzzes the main logics here? The upstream method is already scoped under the api-version condition in the commands.py, so I don't believe this check is ever needed. You can use code comments if there are something important you like to communicate
2. Reuse the same `try-catch` block for both policy definition and definition set.

I would imaging the code would look like:
```python
def _get_custom_or_builtin_policy(client, name, for_policy_set=False):
    from msrest.exceptions import HttpOperationError
    op_groups = client.policy_set_definitions if for_policy_set else client.policy_definitions
    try:
        return op_groups.get(name)
    except (CloudError, HttpOperationError) as ex:
        status_code = ex.status_code if isinstance(ex, CloudError) else ex.response.status_code
        if status_code == 404:
            return op_groups.get_built_in(name)
        raise
```
One question here though, why  `policy_set_definitions` doesn't raise regular `CloudError`, rather its own model error type?

Reply:
@yugangw-msft Thanks for the feedback

1. You are right. We can remove the supported version check. Though, the reason for that is not the scoped logic in commands.py. This PR was addressing the issue with policy assignment create command, but looks like I already scope the set definition params in params.py [here](https://github.com/Azure/azure-cli/blob/master/src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_params.py#L104), so we can indeed remove this.
2. Yes, we can consolidate both try-catch into one. I thought about it, and the reason I didn't want to do it was because it didn't make the code more "readable", but now that you've put this up here, doesn't look bad at all. 👍 I'll push a commit with this.

The reason policy_set_definitions doesn't raise CloudError is because we made a decision, going forward, to add default error object for our [specs](https://github.com/Azure/azure-rest-api-specs/blob/current/specification/resources/resource-manager/Microsoft.Authorization/2017-06-01-preview/policySetDefinitions.json#L89). This is not the case with already released policy definitions spec, so we have to resort to catching different exceptions here.